daemon/mqtt.c

```diff
@@ -0,0 +1,401 @@
+        mqtt_recv_maximum  = config_get_number(CONFIG_SECTION_MQTT, "mqtt receive maximum", 20);
+
+        mqtt_broker_hostname = config_get(CONFIG_SECTION_MQTT, "mqtt broker hostname",  "localhost");
+        mqtt_broker_port     = config_get_number(CONFIG_SECTION_MQTT, "mqtt broker port",  1883);
+
+        info("Maximum parallel outgoing messages %d", mqtt_send_maximum);
+        info("Maximum parallel incoming messages %d", mqtt_recv_maximum);
+
+        // This will setup the base publish topic internally
+        get_publish_base_topic(PUBLICH_TOPIC_GET);
+        init = 1;
+    }
+
+    // initialize mosquitto library
+    rc = mosquitto_lib_init();
+    if (unlikely(rc != MOSQ_ERR_SUCCESS)) {
+        error("Failed to initialize MQTT (libmosquitto library)");
+        return 1;
+    }
+
+    mosq = mosquitto_new(NULL, true, NULL);
+    if (unlikely(!mosq)) {
+        mosquitto_lib_cleanup();
+        error("MQTT new structure  -- %s", mosquitto_strerror(errno));
+        return 1;
+    }
+
+    mosquitto_connect_v5_callback_set(mosq, connect_callback);
+
+    mosquitto_disconnect_v5_callback_set(mosq, disconnect_callback);
+
+    rc = mosquitto_threaded_set(mosq,1);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        error("Failed to tune the thread model for libmoquitto (%s)",mosquitto_strerror(rc));
+
+    // attempt V5 protocol for now, no doc on the return status for now
+    rc = mosquitto_int_option(mosq, MQTT_PROTOCOL_V5, 0);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        error("MQTT protocol specification rc = %d (%s)", rc, mosquitto_strerror(rc));
+
+
+    // These added for the V5 -- may need to be removed if we go to 3.x
+    rc = mosquitto_int_option(mosq, MOSQ_OPT_RECEIVE_MAXIMUM, mqtt_recv_maximum);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        error("MQTT receive maximum queue set failed rc = %d (%s)", rc, mosquitto_strerror(rc));
+
+    rc = mosquitto_int_option(mosq, MOSQ_OPT_SEND_MAXIMUM, mqtt_send_maximum);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        error("MQTT send maximum queue set failed rc = %d (%s)", rc, mosquitto_strerror(rc));
+
+    rc = mosquitto_int_option(mosq, MOSQ_OPT_SEND_MAXIMUM, 1);
+    info("MQTT in flight messages set to 1  -- %s",mosquitto_strerror(rc));
+
+    rc = mosquitto_reconnect_delay_set(mosq, NETDATA_MQTT_RECONNECT_DELAY, NETDATA_MQTT_MAX_RECONNECT_DELAY, 1);
+
+    //mosquitto_tls_set(mosq, "/etc/netdata/mqtt/ca.crt", NULL, "/etc/netdata/mqtt/server.crt", "/etc/netdata/mqtt/server.key", NULL);
+
+    rc = mosquitto_connect_async(mosq, mqtt_broker_hostname, mqtt_broker_port, NETDATA_MQTT_PING_INTERVAL);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        error("Connect %s MQTT status = %d (%s)", netdata_configured_hostname, rc, mosquitto_strerror(rc));
+    else
+        info("Establishing MQTT link to %s", netdata_configured_hostname);
+
+    // TODO: This may need to be elsewhere -- also check return code
+    mqtt_subscribe("netdata/command");
```

Comment:
Only to confirm, because I did not study the library libmosquitto yet, I verified that you were reporting errors in different parts of this function, but you never returned from errors, so  I understood that the library sets default values every time that you moved in front, is my conclusion correct?

Reply:
Several functions will fail becaus of incorrect arguments, others when the actual connection is attempted (e.g. you never get a connection established callback for example)

For the subscribe command more error handling will be added (also it may be moved around) depending on the agent cloud link protocol